mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java

```diff
@@ -0,0 +1,369 @@
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+import javax.mail.Flags;
+import javax.mail.Flags.Flag;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.james.backends.postgres.utils.PostgresExecutor;
+import org.apache.james.mailbox.MessageUid;
+import org.apache.james.mailbox.ModSeq;
+import org.apache.james.mailbox.model.ComposedMessageIdWithMetaData;
+import org.apache.james.mailbox.model.MessageMetaData;
+import org.apache.james.mailbox.model.MessageRange;
+import org.apache.james.mailbox.model.UpdatedFlags;
+import org.apache.james.mailbox.postgres.PostgresMailboxId;
+import org.apache.james.mailbox.postgres.PostgresMessageId;
+import org.apache.james.mailbox.postgres.mail.PostgresMessageModule.MessageTable;
+import org.apache.james.mailbox.store.mail.model.MailboxMessage;
+import org.apache.james.mailbox.store.mail.model.impl.SimpleMailboxMessage;
+import org.jooq.Condition;
+import org.jooq.DSLContext;
+import org.jooq.Record;
+import org.jooq.Record1;
+import org.jooq.SelectFinalStep;
+import org.jooq.SelectSeekStep1;
+import org.jooq.SortField;
+import org.jooq.TableOnConditionStep;
+import org.jooq.UpdateConditionStep;
+import org.jooq.UpdateSetStep;
+import org.jooq.impl.DSL;
+
+import com.google.common.collect.Iterables;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class PostgresMailboxMessageDAO {
+
+    private static final TableOnConditionStep<Record> MESSAGES_JOIN_MAILBOX_MESSAGES_CONDITION_STEP = TABLE_NAME.join(MessageTable.TABLE_NAME)
+        .on(TABLE_FIELD.of(TABLE_NAME, MESSAGE_ID).eq(TABLE_FIELD.of(MessageTable.TABLE_NAME, MessageTable.MESSAGE_ID)));
+
+    public static final SortField<Long> DEFAULT_SORT_ORDER_BY = MESSAGE_UID.asc();
+
+    private static SelectFinalStep<Record1<Long>> selectMessageUidByMailboxIdAndExtraConditionQuery(PostgresMailboxId mailboxId, Condition extraCondition, Optional<Integer> limit, DSLContext dslContext) {
+        SelectSeekStep1<Record1<Long>, Long> queryWithoutLimit = dslContext.select(MESSAGE_UID)
+            .from(TABLE_NAME)
+            .where(MAILBOX_ID.eq((mailboxId.asUuid())))
+            .and(extraCondition)
+            .orderBy(MESSAGE_UID.asc());
+        return limit.map(limitValue -> (SelectFinalStep<Record1<Long>>) queryWithoutLimit.limit(limitValue))
+            .orElse(queryWithoutLimit);
+    }
+
+    private final PostgresExecutor postgresExecutor;
+
+    public PostgresMailboxMessageDAO(PostgresExecutor postgresExecutor) {
+        this.postgresExecutor = postgresExecutor;
+    }
+
+    public Mono<MessageUid> findFirstUnseenMessageUid(PostgresMailboxId mailboxId) {
+        return postgresExecutor.executeRow(dslContext -> Mono.from(selectMessageUidByMailboxIdAndExtraConditionQuery(mailboxId,
+                IS_SEEN.eq(false), Optional.of(1), dslContext)))
```

Comment:
Do you think a Builder for selectMessageUidByMailboxIdAndExtraConditionQuery could help better readability?

Reply:
it will be better readability
IMO it still fine when no builder (with 4 parameters). 
and this is a  private method.

Reply:
Ok with that.